### ISSUE TITLE
Transform from camera frame to mount coordinates in optical error source

### ISSUE THREAD

**Issue by bgottula** [seeing-things/track#52]
The optical error source currently makes the assumption that the x-axis corresponds to azimuth error and the y-axis corresponds to altitude error. Furthermore it assumes that this mapping is a linear relationship. This approximation is good at low altitudes but breaks down at higher altitudes. I have noticed that it can take a long time for azimuth to converge when using optical tracking on stars at higher altitudes. I believe this is because the azimuth error can be fairly large even though the star is not very far from the center of the frame in the x-axis. The error signal is smaller than it should be, and thus the control loop takes a longer time to converge than it should.

**Comment by bgottula** [comment]
The challenge here is that the proper coordinate transform requires knowledge of the mount's current altitude (or at least approximate knowledge of this). There are a couple of problems with this:

- The optical tracking loop should be able to function without any such knowledge. For example, it should still work even when alignment has not been performed. One option would be to only do the proper coordinate transformation when position knowledge is available and fall back on the current approach when it's not available.
- It takes tens of milliseconds to read it from the mount which is expensive.

**Comment by bgottula** [comment]
This might be somewhat less troublesome with the G11 mount for the following reasons:

- Even without performing alignment, as long as the mount is started up in the recommended position (counter weight down) the mount's knowledge of its position relative to the pole will be pretty decent.
- Reading position from the Gemini 2 is fast, only a few ms

I think this should be implemented in a manner such that it will use absolute position data to do the proper coordinate transform if that information is available, but if not it will fall back on the current approximation.

**Comment by bgottula** [comment]
As of May this was implemented but it was not thoroughly tested and there may still be corner cases that don't work quite right, such as #74 and #86. I think at least a few manual tests of specific cases should be performed to confirm that the error magnitudes in different parts of the sky including near the celestial pole match the expected values.

**Comment by bgottula** [comment]
Since much of the coordinate system code has been rewritten this needs to be redone.

**Comment by bgottula** [comment]
Redone as part of https://github.com/seeing-things/track/pull/154 but not tested at all yet

**Comment by bgottula** [comment]
I did some testing today and found the following:

- Needed to apply a sign change or two in the optical error source. Once this was corrected, a camera rotation value in the mount model parameter set of 0 corresponded to the bottom of the camera (USB connector size, for ZWO) pointing in the same direction as the counterweight shaft.
- Optical error source was not correctly accounting for camera binning in transforming from pixels to degrees
- A couple of times the software paused for nearly a second before continuing. I don't know what caused that, but I suspect it must have been a ZWO camera issue.
- Rotation of the camera by 90 degrees counter-clockwise from 0, when viewed from behind the camera, corresponded to a value of -90 degrees in the mount model parameter set.

**Comment by bgottula** [comment]
Three of the four issues found during recent testing as noted in the previous comment were fixed in https://github.com/seeing-things/track/pull/185.

Outstanding:

- [ ] Find root cause of long pauses
- [ ] Confirm that transformations are correctly scaling the right ascension axis error term as a function of the declination axis position

**Comment by bgottula** [comment]
On 25 April 2020 I performed testing of optical tracking where I had access to the full sky. I noted that I was able to successfully track Venus west of the mount meridian with an appropriate value for the guidescope camera orientation in the mount model parameter set (-90 degrees in my case, which corresponded to the ZWO ASI178MC oriented such that the USB connector was facing to the right when viewed from behind). However, tracking did not work east of the mount meridian for the same value--the target was pushed out of the field of view.

- [ ] Fix / test tracking on both sides of mount meridian

**Comment by bgottula** [comment]
I performed further testing on 26 April 2020 on both sides of the meridian and found optical tracking to be working as expected. This was after refactoring a bit of the code and after fixing a couple of sign changes in the process. Not known for certain whether the issue encountered on the 25th was fixed in the process, but it was working with the code I refactored.

I also did some testing with a manufactured target at a fixed offset in the camera frame to test whether the encoder error terms were scaled appropriately based on the declination axis position, and I found that they were indeed being scaled as I expected. Closing this out. Will open a separate issue for the long pauses if that issue recurs.

**Comment by bgottula** [comment]
Welp. Testing of optical tracking of Venus out in the field on 3 May 2020 was unsuccessful. For that test the mount was cold-started and no alignment was performed, so optical tracking was using a default mount model. While debugging in the field I thought that perhaps the problem was that the mount pole altitude was set to 90 degrees, but editing this to 38 degrees to more closely match the hardware did not solve the problem. Telemetry showed: 
- even though the declination axis was far from the pole, the error terms for the mount axes were very large (dozens of degrees) when neither should have been more than a degree or two
- error direction and magnitude in camera frame looked correct
- mount encoder positions looked believable (about the right difference from the 180-degree startup positions) but these were not checked in detail.

I also tried to rule out how meridian side is handled for optical tracking as a source of the problem by re-starting the program several times while already pointed near Venus. Optical tracking detects the meridian side at startup and continues using this same side from that point forward, so re-starting the program should have detected the correct side.

The evidence suggests that the problem is in the transformation from camera frame to mount encoder positions. There should be sufficient telemetry available to reproduce.

**Comment by bgottula** [comment]
On further investigation I found that I was able to reproduce the large error terms. I traced the root cause to a mistake I made when hand-editing the mount model parameters to account for the orientation of the guide scope camera. Since there was no way to specify this orientation when running `optical_track` without performing an alignment I edited the body of `model.load_default_model()` but instead of modifying `guide_cam_orientation` I mistakenly changed `camera_tilt` from 0 to -90 degrees. Oops! Clearly I need to provide a way to specify this from the command line or config file when `optical_track` is run with the default mount model. Opened #190 to address this.

Closing again with the assumption that the problems observed on 3 May were not the result of a bug in the code.
